core/src/main/scala/org/apache/spark/internal/config/History.scala

```diff
@@ -211,4 +211,11 @@ private[spark] object History {
     .version("3.1.0")
     .bytesConf(ByteUnit.BYTE)
     .createWithDefaultString("2g")
+
+  val HYBRID_STORE_DISK_BACKEND = ConfigBuilder("spark.history.store.hybridStore.diskBackend")
+    .doc("Specifies a disk-based store used in hybrid store; 'leveldb' or 'rocksdb'.")
+    .version("3.3.0")
+    .stringConf
+    .checkValues(Set("leveldb", "rocksdb"))
```

Comment:
Sorry for the late and post-review.

Looks like we tend to transform the value to lowercase only or uppercase only, before applying checks and using it. Preferably we define enum for values of config as well.

https://github.com/apache/spark/blob/7d88f1c5c7f38c0f1a2bd5e3116c668d9cbd98b1/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala#L153-L169

Would you mind if I ask about addressing this as a follow-up PR? Thanks in advance!

Reply:
Sure, I'll do. Thank you for the review.

BTW, after one or two releases, I'm going to propose the removal of `LevelDB` dependency completely. Do you have any concern about the removal of LevelDB JNI, @HeartSaVioR ?

Reply:
Here is the follow-up PR.
- https://github.com/apache/spark/pull/34986

Reply:
>  I'm going to propose the removal of LevelDB dependency completely. Do you have any concern about the removal of LevelDB JNI, @HeartSaVioR ?

Two points I can think of:

1. How I interpret the benchmark in PR description of #34913? Since it takes histogram and uses timer, I'd guess it is about latency, and then looks like RocksDB is "slower" than LevelDB. It's good if I'm mistaken, but if I'm not mistaken, pretty sure I don't support the plan to bring performance regression. "It doesn't work with Apple Silicon" is not sufficient fact to drop it out.

2. How we provide smooth migration from LevelDB to RocksDB? End users already loaded their old applications via LevelDB. This applies to LevelDB KVStore and current Hybrid KVStore backed by LevelDB KVStore.

Reply:
And please propose "deprecation" before "removal", so that everyone is not surprised for the proposal.

Reply:
Thank you for your feedback, @HeartSaVioR .
1. Of course, yes for the deprecation!
2. And, yes, `RocksDB` didn't outperform in my benchmark although I tried several storage optimization before making this PR. So, I put it out from this PR and didn't propose the deprecation yet (so far). The performance tuning is also on my roadmap in Apache Spark 3.3 timeframe.

Reply:
I forgot to reply this part.
> How we provide smooth migration from LevelDB to RocksDB? End users already loaded their old applications via LevelDB. This applies to LevelDB KVStore and current Hybrid KVStore backed by LevelDB KVStore.

It depends on the definition of `smooth` migration. 1) Currently, Spark dropped and rebuild the local db when the corruption happens. We can treat in that way. 2) We can add some copying logic from LevelDB to RocksDB, not vise versa.

BTW, `RocksDB` backend needs to catch up the performance first before discussing the migration. In short, it's too early to consider the migration. Let me re-initiate the discussion when it's ready.

Reply:
Sounds great. Let's defer the discussion later till we get enough rationalization to replace LevelDB with RocksDB.